src/EditorFeatures/Core/Implementation/AutomaticCompletion/Sessions/EditorBraceCompletionSession.cs

```diff
@@ -0,0 +1,84 @@
+        {
+            _braceCompletionService = braceCompletionService;
+        }
+
+        public BraceCompletionResult? GetChangesAfterReturn(IBraceCompletionSession session, CancellationToken cancellationToken)
+        {
+            var context = GetBraceCompletionContextFromSession(session);
+            if (context == null)
+            {
+                return null;
+            }
+
+            return _braceCompletionService.GetTextChangeAfterReturnAsync(context, supportsVirtualSpace: true, cancellationToken).WaitAndGetResult(cancellationToken);
+        }
+
+        public virtual bool AllowOverType(IBraceCompletionSession session, CancellationToken cancellationToken)
+        {
+            var context = GetBraceCompletionContextFromSession(session);
+            if (context != null)
+            {
+                return _braceCompletionService.AllowOverTypeAsync(context, cancellationToken).WaitAndGetResult(cancellationToken);
+            }
+
+            return true;
+        }
+
+        public BraceCompletionResult? GetBraceCompletion(IBraceCompletionSession session, CancellationToken cancellationToken)
+        {
+            var context = GetBraceCompletionContextFromSession(session);
+            if (context == null)
+            {
+                return null;
+            }
+
+            return _braceCompletionService.GetBraceCompletionAsync(context, cancellationToken).WaitAndGetResult(cancellationToken);
+        }
+
+        public BraceCompletionResult? GetChangesAfterCompletion(IBraceCompletionSession session, CancellationToken cancellationToken)
+        {
+            var context = GetBraceCompletionContextFromSession(session);
+            if (context == null)
+            {
+                return null;
+            }
+
+            return _braceCompletionService.GetTextChangesAfterCompletionAsync(context, cancellationToken).WaitAndGetResult(cancellationToken);
+        }
+
+        public bool IsValidForBraceCompletion(char brace, int openingPosition, Document document, CancellationToken cancellationToken)
+            => _braceCompletionService.IsValidForBraceCompletionAsync(brace, openingPosition, document, cancellationToken).WaitAndGetResult(cancellationToken);
+
+        private static BraceCompletionContext? GetBraceCompletionContextFromSession(IBraceCompletionSession session)
+        {
+            var snapshot = session.SubjectBuffer.CurrentSnapshot;
+            var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document != null)
+            {
+                var closingSnapshotPoint = session.ClosingPoint.GetPosition(snapshot);
+                var openingSnapshotPoint = session.OpeningPoint.GetPosition(snapshot);
+                var caretPosition = session.GetCaretPosition()?.Position;
+                return new BraceCompletionContext(document, openingSnapshotPoint, closingSnapshotPoint, caretPosition);
+            }
+
+            return null;
```

Comment:
in general, please flip these checks.  in other words don't do:

```
if (!something)
{
     big complex logic
}

// bail out logic
```

do

```
if (something)
   bail out

// big complex logic
```

this is much easier to grok as the simple case can be understood and validated immediately, before finally getting to the 'meat'.  with the above form, one has to examine the complex case, then go back and check the condition again to see if it fits the bail out case.